src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+package org.apache.logging.log4j.ext.kusto;
+
+import org.apache.http.HttpHost;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.status.StatusLogger;
+
+import com.microsoft.azure.kusto.data.HttpClientProperties;
+import com.microsoft.azure.kusto.data.auth.ConnectionStringBuilder;
+import com.microsoft.azure.kusto.data.exceptions.KustoDataExceptionBase;
+import com.microsoft.azure.kusto.ingest.IngestClient;
+import com.microsoft.azure.kusto.ingest.IngestClientFactory;
+import com.microsoft.azure.kusto.ingest.IngestionMapping;
+import com.microsoft.azure.kusto.ingest.IngestionProperties;
+import com.microsoft.azure.kusto.ingest.exceptions.IngestionClientException;
+import com.microsoft.azure.kusto.ingest.exceptions.IngestionServiceException;
+import com.microsoft.azure.kusto.ingest.result.IngestionResult;
+import com.microsoft.azure.kusto.ingest.source.StreamSourceInfo;
+
+import static com.microsoft.azure.kusto.ingest.IngestionMapping.IngestionMappingKind.CSV;
+import static com.microsoft.azure.kusto.ingest.IngestionMapping.IngestionMappingKind.JSON;
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+import static org.apache.logging.log4j.ext.kusto.Constants.INGESTION_RETRIES;
+
+import io.github.resilience4j.core.IntervalFunction;
+import io.github.resilience4j.retry.Retry;
+import io.github.resilience4j.retry.RetryConfig;
+import io.github.resilience4j.retry.RetryRegistry;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Properties;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A singleton class that does all the work on the Ingestion of data. Is a singleton instance that instantiates the Ingest client
+ * and all operations that are performed from the perspective of ingestion
+ * Uses an external library for performing queued ingestion retries and falls back in case of retries exhausted to write data to a backout
+ * directory where the data can be uploaded in case of transient failures manually.
+ * To avoid configuration where too many retries can be configured , the retry attempts have been "hard-limited" (opinionated view) at 3

```

Comment:
```suggestion
 * To avoid configuration where too many retries can be configured, the retry attempts have been "hard-limited" (opinionated view) at 3
```

Reply:
Also, what do you mean by, "opinionated view"?